bigflow/configuration.py

```diff
@@ -27,43 +22,64 @@ def __init__(self,
         self.add_configuration(name, properties, is_default)
 
     def __str__(self):
-        return "\n".join(map(self.pretty_print, self.config.keys()))
+        return "".join(map(self.pretty_print, self.config.keys())).rstrip("\n")
 
     def resolve_property(self, property_name: str, env_name: str = None):
-        static_props = self.resolve(env_name)
-
-        if property_name in static_props:
-            return static_props[property_name]
-
-        return self._resolve_property_from_os_env(property_name)
+        try:
+            return self.resolve(env_name)[property_name]
+        except KeyError:
+            raise ValueError(f"Failed to load property '{property_name}' from config, also there is no 'bf_{property_name}' env variable.")
 
     def pretty_print(self, env_name: str = None):
         s = io.StringIO()
         pp = pprint.PrettyPrinter(indent=4, stream=s)
+        _, env_name = self._get_env_config(env_name)
 
-        env_config = self._get_env_config(env_name)
-        s.write(env_config.name)
+        s.write(env_name)
         s.write(" config:\n")
         pp.pprint(self.resolve(env_name))
 
-        return s.getvalue()[:-1]
+        return s.getvalue()
+
+    @staticmethod
+    def _capture_osenv_properties():
+        return {
+            k[3:]: v
+            for k, v in os.environ.items()
+            if k.startswith("bf_")
+            and k != 'bf_env'
+        }
 
     def resolve(self, env_name: str = None) -> dict:
-        env_config = self._get_env_config(env_name)
+        env_config, env_name = self._get_env_config(env_name)
+
+        properties_with_placeholders = dict(env_config)
+        for k, v in self._capture_osenv_properties().items():
+            if properties_with_placeholders.get(k, None) is None:
+                properties_with_placeholders[k] = v
+        properties_with_placeholders.setdefault('env', env_name)
+
+        for k, v in properties_with_placeholders.items():
+            if v is None:
+                raise ValueError(f"Failed to load property '{k}' from OS environment, no such env variable: 'bf_{k}'.")
 
-        properties_with_placeholders = {key: self._resolve_property_with_os_env_fallback(key, value)
-                                        for (key, value) in env_config.properties.items()}
+        res = {
+            key: self._resolve_placeholders(value, properties_with_placeholders)
+            for key, value in properties_with_placeholders.items()
+        }
 
-        string_properties = {k: v for (k, v) in properties_with_placeholders.items() if isinstance(v, str)}
-        string_properties['env'] = env_config.name
+        if 'env' not in env_config:
+            # For backward compatability - don't show "magic" 'env' variable to user
+            res.pop('env', None)
```

Comment:
It's a useful little thing (we use it in our projects).

Reply:
It is still here (so might be used during interpolation), but excluded from '.resolve()'
I doubted about this, so probably we should not overcomplicate this and always add 'env' property with current config name (even when it is not specified via os.environ).
wdyt?

Reply:
Sounds like a consistent solution, agree